src/core/etl/src/Flow/ETL/Function/ConcatWithSeparator.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Flow\ETL\Function;
+
+use function Flow\ETL\DSL\type_string;
+use Flow\ETL\PHP\Type\Caster;
+use Flow\ETL\Row;
+
+final class ConcatWithSeparator extends ScalarFunctionChain
+{
+    /**
+     * @var array<ScalarFunction|string>
+     */
+    private readonly array $refs;
+
+    public function __construct(
+        private readonly ScalarFunction|string $separator,
+        ScalarFunction|string ...$refs,
+    ) {
+        $this->refs = $refs;
+    }
+
+    public function eval(Row $row) : mixed
+    {
+        $separator = (new Parameter($this->separator))->asString($row);
+
+        if (!\is_string($separator)) {
+            return '';
+        }
+
+        $values = \array_map(fn (ScalarFunction|string $string) : mixed => \is_string($string) ? $string : Caster::default()->to(type_string(true))->value($string->eval($row)), $this->refs);
```

Comment:
Instead of map, it could be merged into below loop, so it won't be doing two loops over given values.

Reply:
https://github.com/flow-php/flow/pull/1393